lib/syscall_shim/uk_prsyscall.c

```diff
@@ -0,0 +1,1089 @@
+			uk_streambuf_strcpy(sb, "NULL");
+			uk_streambuf_shcc(sb, fmtf, RESET);
+		} else {
+			unsigned int left;
+			char *c;
+
+			uk_streambuf_shcc(sb, fmtf, CHARP);
+			uk_streambuf_strcpy(sb, "\"");
+			for (c = (char *) param,
+				left = MIN((int) arg, __PR_BUFP_MAXLEN);
+				left > 0;
+				left--, c++) {
+				if (*c >= ' ' && *c <= '~')
+					uk_streambuf_printf(sb, "%c", *c);
+				else
+					uk_streambuf_printf(sb, "\\x%02X",
+							    (int) *c);
+			}
+			uk_streambuf_strcpy(sb, "\"");
+			uk_streambuf_shcc(sb, fmtf, RESET);
+			if (arg > __PR_BUFP_MAXLEN)
+				uk_streambuf_strcpy(sb, "...");
+		}
+		break;
+	case PT_FD:
+		param_fd(sb, fmtf, param);
+		break;
+	case PT_DIRFD:
+		param_dirfd(sb, fmtf, param);
+		break;
+	case PT_PID:
+		param_pid(sb, fmtf, param);
+		break;
+	case PT_OFLAGS:
+		param_oflags(sb, fmtf, param);
+		break;
+#if CONFIG_LIBPOSIX_MMAP || CONFIG_LIBUKMMAP
+	case PT_PROTFLAGS:
+		param_protflags(sb, fmtf, param);
+		break;
+	case PT_MAPFLAGS:
+		param_mapflags(sb, fmtf, param);
+		break;
+#endif /* CONFIG_LIBPOSIX_MMAP || CONFIG_LIBUKMMAP */
+#if CONFIG_LIBPOSIX_FUTEX
+	case PT_FUTEXOP:
+		param_futexop(sb, fmtf, param);
+		break;
+#endif /* CONFIG_POSIX_FUTEX */
+	case PT_CLOCKID:
+		param_clockid(sb, fmtf, param);
+		break;
+#if CONFIG_LIBPOSIX_SOCKET
+	case PT_SOCKETAF:
+		param_socketaf(sb, fmtf, param);
+		break;
+	case PT_SOCKETTYPE:
+		param_sockettype(sb, fmtf, param);
+		break;
+	case PT_MSGFLAGS:
+		param_msgflags(sb, fmtf, param);
+		break;
+#endif /* CONFIG_LIBPOSIX_SOCKET */
+#if CONFIG_LIBPOSIX_PROCESS_CLONE
+	case PT_CLONEFLAGS:
+		param_cloneflags(sb, fmtf, param);
+		break;
+#endif /* CONFIG_LIBPOSIX_PROCESS_CLONE */
+	default:
+		uk_streambuf_shcc(sb, fmtf, VALUE);
+		uk_streambuf_printf(sb, "0x%lx", (unsigned long) param);
+		uk_streambuf_shcc(sb, fmtf, RESET);
+		break;
+	}
+
+out:
+	return;
+}
+
+/* Pretty print a system call return code */
+static void pr_retcode(struct uk_streambuf *sb, int fmtf,
+		       enum param_type type, long retval)
+{
+	UK_ASSERT(sb);
+	UK_ASSERT(uk_streambuf_buf(sb));
+
+	if (PTRISERR((void *) retval) && (retval != 0)) {
+		uk_streambuf_shcc(sb, fmtf, ERROR);
+		uk_streambuf_printf(sb, "%s (%ld)", strerror((int) -retval),
+				    retval);
+		uk_streambuf_shcc(sb, fmtf, RESET);
+	} else if (type == PT_BOOL) {
+		/* Replace boolean with OK (here, errors are already catched) */
+		uk_streambuf_shcc(sb, fmtf, OK);
+		uk_streambuf_strcpy(sb, "OK");
+		uk_streambuf_shcc(sb, fmtf, RESET);
+	} else {
+		/* Use same formatting as for parameters */
+		pr_param(sb, fmtf, type, retval);
+	}
+}
+
+/*
+ * Helper macros for pretty print a system call request (PR_SYSCALL,
+ * VPR_SYSCALL) and to pretty print the return code of a system call
+ * (PR_SYSRET).
+ */
+#define __PR_SYSCALL_PARAM_PRINT(i, sb, type, arg)			\
+	do {								\
+		if ((i) != 0)						\
+			uk_streambuf_strcpy((sb), ", ");		\
+		pr_param((sb), fmtf, (type), (long) (arg));		\
+	} while (0)
```

Comment:
@skuenzer, I know this is weird, but it seems that after removing the `semicolon` here after `while (0)`, the unikernel just doesn't compile anymore. Adding the semicolon back seems to solve this, but I'm guessing the problem may actually be from somewhere else (?)

Reply:
ah, right - I wrongly fixed it, the semicolon must be there because of the VARG_EXPAND macro. I need to add a checkpatch ignore then, it is not prepared for such complex macro expansions.